### ISSUE TITLE
Endless recursive function call when mouse selecting inner cells

### ISSUE THREAD

**Issue by gunterkoenigsmann** [wxMaxima-developers/wxmaxima#1387]
How to reproduce:

     a/b^5

...and try to select the "a", "b" or "5".

backtrace:

````
#1336 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1337 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1338 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1339 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1340 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1341 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
--Type <RET> for more, q to quit, c to continue without paging--
#1342 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1343 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1344 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1345 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1346 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1347 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1348 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1349 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:1013
#1350 0x0000555555841d5d in Cell::GetCellsInRect(wxRect const&) const
    (this=0x555556b1ef40, rect=...)
    at /home/gunter/src/wxmaxima/src/cells/Cell.cpp:988
#1351 0x0000555555841ff5 in Cell::GetInnerCellsInRect(wxRect const&) const
````

**Comment by KubaO** [comment]
I probably broke it with a recent commit. I have an idea what events to add to be able to have test cases for mouse-selecting cells. I’d say we delay 20.08, maybe skip it and have a better 20.09. My timeframe to fix these things is about a week from now - I am very busy with other work.

**Comment by gunterkoenigsmann** [comment]
Will delay the release then and try to get rid of another per-cell variable in the meantime.

---

The mouse selection code looks really cool and minimalistic. I guess that is the kind of thing you meant when you wrote you can half the amount of code wxMaxima consists of.

**Comment by KubaO** [comment]
As long as I can make it not buggy :) The existing code in general has a lot of corner cases and assumptions and sometimes I miss them when refactoring.

There will be even bigger cuts in the coming months, and those changes that now break things are being factored out from the big change I work on. I usually prototype things that break stuff but indicate a possible direction of development, and then re-develop the individual steps that lead to the intention of the big prototype. The finished code ends up often looking nothing like the prototype, but that’s OK - it is usually better that way.

The big change coming is the introduction of CellContext that will be used as an intermediary when operating on cells. It will be the owner of a group cell list. Anyone wanting to use cells will be using the context object and the cell operations will not be directly available (they will be protected and available only to the friend CellContext). Right now we pay a big price: every cell carries the configuration pointer (and sometimes the cellpointers pointer too), and various “old_” values just to detect if something external changed that would require a recalculation. Instead, the cell context object will keep this information, and will maintain ranges of cells that are “dirty”, as well as keeping other cell pointers, and the cells will not store any pointers to the context.

The context will keep implicit cell lists, such that the cells won’t even have the next and previous pointers! They’ll be laid out contiguously in memory (each cell being still of arbitrary size), and the context will be the one providing iterators for such lists. The group pointer will also be gone, since the cells won’t need to know: the context will of course have this information and will thus provide it to cells whenever calling into them. 

In a way, a lot of functionality that’s now in the base Cell class will move out to CellContext - the goal will be to make cells themselves dumber, since they can’t really see the big picture, whereas the CellContext will always know what is the operation being carried out at the moment. Lots of cell and worksheet code is now strewn with very convoluted code to deduce such state. There’s code buried within cells that looks at their predecessor or successor - such code doesn’t logically belong in a cell, but in a higher layer that manages the cell lists. 

The cell methods will be passed the cell context as the first argument. Subsequently, the operations that touch cellpointers will move to the context object and at that point the CellPtr class will be deleted (my goal really was not to add more code!) and the Observed base class will vanish from the cell, removing another pointer.

A few more changes are planned that will then bring the base Cell class below one cacheline in size - that’s less than 64 bytes. We’re now at the 120 byte size, and originally it was ~340 bytes or so. My bold and unstated plan was to cut Cell size by an order of magnitude vs the size at the beginning of 2020, and this will indeed happen - it will be in the 32-48 byte range. The derived classes will follow suit and their overhead will be drastically reduced as well.

A basic “dumb” TextCell should not really be any bigger than one cacheline, and indeed we’ll implement a clang-style string class that will be 16 bytes long and will be able to store up to 15 characters as a small-string optimization. That string will be the only field added on top of what’s in the Cell, bringing it’s size to exactly one cacheline. 

The Cell will not have any C++ `virtual` methods anymore - instead, the virtual method table pointer will be selected by a byte-sized index into a vector of method tables. We’ll do our own virtual method dispatch that way to save memory, and to have our own RTTI (runtime type information) without C++’s unknown overheads (the standard guarantees nothing as far as “native” RTTI goes, really).

To recap: due to the CellContext introduction, the Cell will lose:
- vtable pointer (8 bytes)
- Observed (8 bytes)
- prev/next pointers (16 bytes)
- group pointer (8 bytes)
- configuration pointer (8 bytes)
- accessible pointer (8 bytes)
- tooltip pointer (8 bytes) to be replaced by a 2-byte tooltip index
- m_clientwidth_old (4 bytes)
- all booleans that indicate whether anything needs to be recalculated

Together, that’s a 66 byte reduction, down to 54 bytes. Further planned removal of fullWidth and lineWidth (to be stored only once per cell list), instead of in each cell, will bring it down to 46 bytes.

Once the recalculation logic will be refactored, further reductions in per-cell size information will be possible.

So that’s roughly what’s in store. When those changes are all done, the “should we recalculate” logic will be trivial and fit on two screens of code, and it will be handled by the cell context on behalf of the cells, since the context will have a full view of what changed externally (eg client width, added cells, etc). The result: timing-invariant recalculation/repaint that won’t be affected by any platform UI event ordering etc.

All those things are very closely intertwined and are way too big to be applied at once - it’d touch thousands of lines of code, and it’s a lot of work to cleanly split it into cohesive pieces that can be each understood. I write my code with educational objective in mind as well, so that each commit is supposed to be comprehensible in isolation, and a reasonable example of improvement/refactoring on a legacy code base. In other words, I’m also using wxMaxima as a demonstrator project that could be potentially used in developer training, bringing new team members on board, etc. It’s not always possible to adhere to such a standard, but for each commit that’s always my starting goal.

**Comment by gunterkoenigsmann** [comment]
Thought about it: Instead of creating an external object that keeps track on the info which cells need to be recalculated - what if each cell contained one int-sized hash of its fontsize and a counter that is stepped each time the configuration changes and each time each cell needs to be recalculated due to whatever reason. Overwriting that hash with an arbitrary value would tell the cell it needs to be recalculated, in that case.

**Comment by KubaO** [comment]
The “info” on what cells need recalculation is basically a couple of iterators. There’s no need to keep any of it in the cells, because the circumstances in which cells become out of date are completely controlled externally and they never originate within the cell. The cell doesn’t randomly decide to become invalid. It’s always the external context that does something that makes a cell’s dimensions and positioning invalid, and only such context is in any position to keep track of it without O(N) memory cost in number of cells.

For example, at the moment the output cells are static once they are created, until the width of the canvas or the zoom or screen dpi is changed. In other words: an output list of cells is invalid while being parsed and created, then it gets recalculated once and it’s done and over. No need to track anything at any moment: before the recalculation loop is done, cells are invalid, once done they become valid, and that’s all. This is just a rather simple example.

But in all cases the context drives what’s happening - for example the context is the one that makes the cells draw themselves and it knows that it needs to recalculate the intersection of the dirty ranges and the drawn ranges before they are drawn - and such intersection can morph as cell positions get updated. The cells will never do anything to themselves - they will never attempt to recalculate or draw themselves. All the list operations (XyzList inside of Cell right now) belong in the context since the context will be the only object with knowledge of presence of cell lists. The cells themselves will be completely solitary with no pointers to anything - not even to internal cells. They’ll know that they have internal cells, but the knowledge of where these cells are in memory will rest with the context since it can derive this information without the need to actually store pointers that would be per-cell. The cell will know how many internal cells it had (the per-cell information), and the context will provide the cell with the location of the cells on request. The cells don’t need to store a number: the internal cell count pretty much depends on the “am I broken into lines” bit flag that’s already there, and the cell only needs a virtual method that gives that number to anyone who asks - typically that would be the context.

The memory management within the context will allocate each group cell in a contiguous memory area with all the inner cells, so that the overall number of allocations will drop quite a bit - to just one per group cell, plus however many strings in the text cells would have more than 15 UTF-8 code units. For quite a bit of math, text cells will not allocate heap for strings! Then there are still tooltips, but the situation there is not tragic: a lot of them can be represented as indexes into a small array of preset tooltips. The rest will indeed need to be heap allocated, at a nominal cost.
